chemprop/models/ffn.py

```diff
@@ -0,0 +1,439 @@
+        :param output_size: The size of output.
+        :param dropout: Dropout probability.
+        :param activation: Activation function.
+        :param atom_constraints: A list of booleans indicating whether constraints applied to output of atomic properties.
+        :param bond_constraints: A list of booleans indicating whether constraints applied to output of bond properties.
+        :param shared_ffn: Whether to share weights in the ffn between different atom tasks and bond tasks.
+        :param weights_ffn_num_layers: Number of layers in FFN for determining weights used to correct the constrained targets.
+        """
+        super().__init__()
+        self.shared_ffn = shared_ffn
+
+        if num_layers > 1 and shared_ffn:
+            self.atom_ffn_base = nn.Sequential(
+                build_ffn(
+                    first_linear_dim=atom_features_size,
+                    hidden_size=atom_hidden_size,
+                    num_layers=num_layers - 1,
+                    output_size=atom_hidden_size,
+                    dropout=dropout,
+                    activation=activation,
+                ),
+                activation,
+            )
+            self.bond_ffn_base = nn.Sequential(
+                build_ffn(
+                    first_linear_dim=2*bond_features_size,
+                    hidden_size=bond_hidden_size,
+                    num_layers=num_layers - 1,
+                    output_size=bond_hidden_size,
+                    dropout=dropout,
+                    activation=activation,
+                ),
+                activation,
+            )
+        else:
+            self.atom_ffn_base = nn.Identity()
+            self.bond_ffn_base = nn.Identity()
+
+        ffn_list = []
+        for constraint in atom_constraints:
+            if constraint:
+                ffn_list.append(
+                    FFNAtten(
+                        features_size=atom_features_size,
+                        hidden_size=atom_hidden_size,
+                        num_layers=num_layers,
+                        output_size=output_size,
+                        dropout=dropout,
+                        activation=activation,
+                        ffn_base=self.atom_ffn_base,
+                        ffn_type="atom",
+                        shared_ffn=shared_ffn,
+                        weights_ffn_num_layers=weights_ffn_num_layers,
+                    )
+                )
+            else:
+                ffn_list.append(
+                    FFN(
+                        features_size=atom_features_size,
+                        hidden_size=atom_hidden_size,
+                        num_layers=num_layers,
+                        output_size=output_size,
+                        dropout=dropout,
+                        activation=activation,
+                        ffn_base=self.atom_ffn_base,
+                        ffn_type="atom",
+                        shared_ffn=shared_ffn,
```

Comment:
both of these objects are parameterized very similarly. Can you pull the common arguments out into some parameter dict `ffn_params` and then pass it in using dictionary unpacking (`**ffn_params`)

Reply:
Sounds good! I should do that.